beacon_node/beacon_chain/src/shuffling_cache.rs

```diff
@@ -127,13 +140,55 @@ impl ShufflingCache {
             // worth two in the promise-bush!
             .map_or(true, CacheItem::is_promise)
         {
-            self.cache.put(
+            self.insert_cache_item(
                 key,
                 CacheItem::Committee(committee_cache.to_arc_committee_cache()),
             );
         }
     }
 
+    /// Prunes the cache first before inserting a new cache item.
+    fn insert_cache_item(&mut self, key: AttestationShufflingId, cache_item: CacheItem) {
+        self.prune_cache();
+        self.cache.insert(key, cache_item);
+    }
+
+    /// Prunes the `cache` to keep the size below the `cache_size` limit, based on the following
+    /// preferences:
+    /// - Entries from more recent epochs are preferred over older ones.
+    /// - Entries with shuffling ids matching the head's previous, current, and future epochs must
+    ///   not be pruned.
+    fn prune_cache(&mut self) {
+        if self.cache.len() >= self.cache_size {
+            let prune_count = self.cache.len() - self.cache_size + 1;
```

Comment:
I think we might have some under-flowing math here. I'm not quite following the math here, but I'd expect something like this to work and be panic-free in the eyes of the compiler:

```rust
let target_cache_size = self.cache_size.saturating_sub(1);
if let Some(prune_count) = self.cache.len().checked_sub(target_cache_size) {
    // Prune.
}
```

Reply:
Updated in f6ef0af03fabad65244208d6851595732ac9fe1e, thanks!